### PR TITLE
number like categories respect `layout.autotypenumbers`

### PR DESCRIPTION
**Pull request by nicholas-esterer**
Fixes issue #5767 
The gist of the fix is in `getRangePosition` in `set_convert.js`.
But some discussion is still required:
The interpreting of category names that are numeric strings should follow what `layout.axis.autotypenumbers` says to do but it is still not clear to me why the trace data is and was always drawn correctly but the shape data was not.

### CODE REVIEW

**Comment by nicholas-esterer** [comment]
Basically what I am asking is: what difference does `layout.axis.autotypenumbers` make to drawing the shape on a category axis? If `autotypenmbers='strict'` the categories haven't been converted to numbers if `autotypenumbers='convert types'` they are strings, but they are still just keys in a dictionary that look up their position. For example, if `xaxis.categoryarray=['10','40','20']` or`xaxis.categoryarray=[10,40,20]` you can still never place a shape at `{x0:10.5 etc.}`, it just draws that part of the shape in the middle of the plot, which is what happens when it doesn't find the requested category.

**Comment by nicholas-esterer** [comment]
Is their a case for supporting what happens currently when the coordinates are numbers?
The current behaviour is: when shapes are drawn on `category` axes and the coordinates look like numbers (i.e., they are numbers or numeric strings) then the coordinates are interpreted as indices into `categoryarray` and the points are placed at the categories in those indices.
Note that this behaviour is NOT seen for traces: if the coordinates look like numbers, the matching category is found, be it number or string (i.e., coordinate 1 can match '1'). If the category is not found, it is added to `categoryarray` at the end.
The question is: for shapes (and other annotations) do we support the indexing behaviour if the coordinates are numbers but the category-matching behaviour if the coordinates are strings? The question is asked because maybe people rely on the old functionality.

**Comment by nicholas-esterer** [comment]
@alexcjohnson @nicolaskruchten

**Comment by alexcjohnson** [comment]
Yes, I think we need to continue treating actual numbers as indices, but treat numeric strings as categories. The primary reason for this is fractional indices, if you want to draw a line halfway between two categories for example.

Traces don't need this, because the whole point of a categorical trace is that the data points are all positioned on categories.

File: src/plots/cartesian/set_convert.js [hunk 1]
```diff
@@ -178,8 +179,8 @@ module.exports = function setConvert(ax, fullLayout) {
         if(isNumeric(v)) return +v;
     }
 
-    function getRangePosition(v) {
-        return isNumeric(v) ? +v : getCategoryIndex(v);
+    function getRangePositionForCategory(v) {
+        return (typeof v !== 'string') ? v : getCategoryIndex(v);
```
Comment (archmoj) [inline src/plots/cartesian/set_convert.js:183]: Could you please elaborate why using `isNumeric` here is not possible?

File: test/image/mocks/annotations-with-numeric-string-categories.json [hunk 2]
```diff
@@ -0,0 +1,57 @@
+{
+"data":[{
+                "x": ["3", "1", "5", "2", "4", "6"],
+                "y": ["6","7","8","2","3","4"]
+            }],
+"layout":{
+                "xaxis": {
+                    "type": "category",
```
Comment (archmoj) [inline test/image/mocks/annotations-with-numeric-string-categories.json:8]: Let's adjust formatting for these new mocks.

File: test/jasmine/assets/numeric_category.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
+'use strict';
```
Comment (archmoj) [inline test/jasmine/assets/numeric_category.js:1]: Let's move all this into `test/jasmine/tests/shapes_test.js`.

File: test/image/mocks/annotations-with-numeric-string-categories.json [hunk 4]
```diff
@@ -0,0 +1,57 @@
+{
+"data":[{
+                "x": ["3", "1", "5", "2", "4", "6"],
+                "y": ["6","7","8","2","3","4"]
```
Comment (archmoj) [inline test/image/mocks/annotations-with-numeric-string-categories.json:4]: What happens if we use numbers here?

File: src/plots/cartesian/set_convert.js (hunk 1 above)
Comment (nicholas-esterer) [reply src/plots/cartesian/set_convert.js:183]: `isNumeric` returns `true` for `isNumeric('1')` which we don't want. We want numeric string categories to have to use the indices.

File: test/image/mocks/annotations-with-numeric-string-categories.json (hunk 4 above)
Comment (nicholas-esterer) [reply test/image/mocks/annotations-with-numeric-string-categories.json]: For this example, it looks the same, but if you try using numbers in the original issue, you get
bugstillthere

which means the coordinates in a shape refer to the index of the category on the axis if the shape coordinates are a number and the referenced axis is categorical. In this case the axes are categorical but the categories are numbers (see how they are spaced evenly on the axis even though the numbers are not evenly spaced)
This is probably surprising to users, but the most surprising thing, which is reported in the referenced issue, is that a numeric string is interpreted as an index into the axis.

For context, this is the code:
```html
<!DOCTYPE html>
<html>
<head>
    <meta http-equiv='Content-Type' content='text/html:charset=utf-8' />
    <script src="../../build/plotly.js"></script>
</head>
<body>
<div id="tester"></div>
<script>
    var layout = {
      "xaxis": {
        "type": "category",
        "categoryarray": [
          10, 25, 30, 40, 50, 60, 70
        ],
      },
      "shapes": [
        {
          "type": "rect",
          "xref": "x",
          "yref": "y",
          "x0": 40,
          "x1": 60,
          "y0": 10,
          "y1": 20,
          "fillcolor": "rgba(45,180,135,0.3)",
        }
      ]
    };

    var data = [
      {
        "x": [ 10, 30, 50, 70],
        "y": [ 10, 20, 30, 40],
      },
    ];
    
    TESTER = document.getElementById('tester');
    Plotly.newPlot( TESTER, data, layout );
</script>
</body>

</html>
```

**Comment by gvwilson** [comment]
This pull request has been sitting for a while, so I would like to close it as part of our effort to tidy up our public repositories. I've assigned it to myself to keep track of it; I'll wait until 2024-06-17 for someone to say it's still relevant and they'll to take it on, and otherwise I will close it then. Thanks - @gvwilson

**Pull request closed without merging**